depthai_ros_driver/include/depthai_ros_driver/impl/ros_agnostic.hpp

```diff
@@ -0,0 +1,249 @@
+      srv_name,
+      [callback = std::move(callback)](const RequestMsg& req, ResponseMsg& res)
+      {
+        // internally convert ref to ptr
+        auto res_ptr = std::make_shared<ResponseMsg>();
+        callback(std::make_shared<RequestMsg>(req), res_ptr);
+        res = *res_ptr;
+        return true;
+      });
+    #endif
+  }
+
+  #if defined(USE_ROS2)
+  std::shared_ptr<void> _ptr;
+  #else
+  ros::ServiceServer _srv;
+  #endif
+
+  friend class NodeInterface;
+};
+
+//==============================================================================
+class Timer
+{
+private:
+  /// @brief Create ros timer
+  template<typename Callback>
+  void create_timer(
+    ROSNodeHandle node_handle,
+    const double period_sec,
+    const Callback& callback)
+  {
+    #if defined(USE_ROS2)
+    const auto period = std::chrono::duration<double>(period_sec);
+    _timer = node_handle->create_wall_timer(period, callback);
+    #else
+    _timer = node_handle->createTimer(ros::Duration(period_sec),
+        [callback = std::move(callback)](const ros::TimerEvent&)
+        {
+          callback();
+        });
+    #endif
+  }
+  #if defined(USE_ROS2)
+  rclcpp::TimerBase::SharedPtr _timer;
+  #else
+  ros::Timer _timer;
+  #endif
+
+  friend class NodeInterface;
+};
+
+//==============================================================================
+template<class Msg>
+Publisher NodeInterface::create_publisher(
+  const std::string& topic_name,
+  const uint32_t queue_size)
+{
+  Publisher pub;
+  pub.create_publisher<Msg>(_nh, topic_name, queue_size);
+  return pub;
+}
+
+//==============================================================================
+template<class Msg, typename Callback>
+Subscription NodeInterface::create_subscription(
+  const std::string& topic_name,
+  const uint32_t queue_size,
+  const Callback& callback)
+{
+  Subscription sub;
+  sub.create_subscription<Msg>(_nh, topic_name, queue_size, callback);
+  return sub;
+}
+
+//==============================================================================
+template<class Msg, typename Callback>
+Service NodeInterface::create_service(
+  const std::string& srv_name,
+  const Callback& callback)
+{
+  Service srv;
+  srv.create_service<Msg>(_nh, srv_name, callback);
+  return srv;
+}
+
+//==============================================================================
+template<typename Callback>
+Timer NodeInterface::create_timer(
+  const double period_sec,
+  const Callback& callback)
+{
+  Timer timer;
+  timer.create_timer(_nh, period_sec, callback);
+  return timer;
+}
+
+//==============================================================================
+const RosTime NodeInterface::current_time()
+{
+  #if defined(USE_ROS2)
+  return _nh->now();
+  #else
+  return RosTime::now();
+  #endif
+}
+
+//==============================================================================
+template<typename Param>
+void get_param(
```

Comment:
Could we get an example of dyn-param support? I know it's not used in the depthai_ros, but it is quite common outside of the package

Reply:
We'll take a look and see if we can come up with something easily. Parameters in ROS 2 are all dynamic, so that's not a show-stopper. The problem is the different APIs in ROS 1 combined with ROS 2 having a whole new API of its own.

Reply:
Let's skip this then. We might use ddynamic parameter here instead which seems to have a similar interface of registering parameters